[GStreamer] Use integers as track ID instead of AtomString

vivienne-w opened the pull request:
#### 7a7367fb8f432c911a00f234ec49949324ca69ef
<pre>
[GStreamer] Use integers as track ID instead of AtomString
<a href="https://bugs.webkit.org/show_bug.cgi?id=270638">https://bugs.webkit.org/show_bug.cgi?id=270638</a>

Reviewed by Philippe Normand.

Switch to using integers for track IDs internally instead of AtomString.

Some notable special cases:
- as MediaStream tracks are recommended by the spec to use UUID strings,
  we just truncate it to 64 bits for the internal TrackID,
  and return the full ID as String (stored in m_gstStreamId) in `trackUID()`
- In MSE, there is the potential for multiple SourceBuffers providing tracks
  with identical IDs. Previously, this was less of an issue since IDs were
  also based on track type, so two tracks &quot;A0&quot; and &quot;V0&quot; would still have
  distinct IDs. However, now they would both get ID 0, and one would get
  dropped in `MediaSourceTrackGStreamer::filterOutRepeatingTracks`
  To work around this, AppendPipeline will register any IDs it creates
  with `MediaSourcePrivateGStreamer::m_trackIdRegistry`, which will
  assign different IDs in case of a collision.
  There has been some consideration as to where to place this registry,
  and how to assign fallback IDs, which is both explained in comments.
  One test that produced these collisions is:
  imported/w3c/web-platform-tests/encrypted-media/clearkey-mp4-playback-temporary-waitingforkey.https.html

* Source/WebCore/platform/graphics/gstreamer/AudioTrackPrivateGStreamer.cpp:
(WebCore::AudioTrackPrivateGStreamer::AudioTrackPrivateGStreamer):
Add constructor with an explicit TrackID argument.
(WebCore::AudioTrackPrivateGStreamer::capsChanged):
* Source/WebCore/platform/graphics/gstreamer/AudioTrackPrivateGStreamer.h:
(WebCore::AudioTrackPrivateGStreamer::create):

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::getTrackIdFromPad): wrapper around gst_pad_get_stream_id, parses stream-id into a TrackID
(WebCore::getTrackIdFromStream): ditto, wraps around gst_stream_get_stream_id
(WebCore::trackIdFromString): parses a stream-id.
Can handle plain integers, the qtdemux-style &quot;.../&lt;ID&gt;&quot; format, and UUIDs.
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:

* Source/WebCore/platform/graphics/gstreamer/InbandTextTrackPrivateGStreamer.cpp:
(WebCore::InbandTextTrackPrivateGStreamer::InbandTextTrackPrivateGStreamer):
Add constructor with an explicit TrackID argument.
* Source/WebCore/platform/graphics/gstreamer/InbandTextTrackPrivateGStreamer.h:
(WebCore::InbandTextTrackPrivateGStreamer::create):

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::notifyPlayerOfTrack):
(WebCore::MediaPlayerPrivateGStreamer::handleTextSample):
(WebCore::MediaPlayerPrivateGStreamer::updateEnabledVideoTrack):
(WebCore::MediaPlayerPrivateGStreamer::updateEnabledAudioTrack):
(WebCore::MediaPlayerPrivateGStreamer::playbin3SendSelectStreamsIfAppropriate):
(WebCore::MediaPlayerPrivateGStreamer::updateTracks):
(WebCore::MediaPlayerPrivateGStreamer::setupCodecProbe):
(WebCore::MediaPlayerPrivateGStreamer::codecForStreamId):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:

* Source/WebCore/platform/graphics/gstreamer/TextSinkGStreamer.cpp:
(webkitTextSinkHandleSample):

* Source/WebCore/platform/graphics/gstreamer/TrackPrivateBaseGStreamer.cpp:
(WebCore::TrackPrivateBaseGStreamer::TrackPrivateBaseGStreamer):
Add constructor with an explicit TrackID argument.
Sets the new m_shouldUsePadStreamId field false, to avoid the ID being changed elsewhere.
(WebCore::TrackPrivateBaseGStreamer::setPad):
(WebCore::TrackPrivateBaseGStreamer::notifyTrackOfStreamChanged):
(WebCore::TrackPrivateBaseGStreamer::installUpdateConfigurationHandlers):
(WebCore::TrackPrivateBaseGStreamer::updateTrackIDFromTags):
(WebCore::TrackPrivateBaseGStreamer::prefixForType): Deleted.
(WebCore::trimStreamId): Deleted.
(WebCore::TrackPrivateBaseGStreamer::generateUniquePlaybin2StreamID): Deleted.
(WebCore::TrackPrivateBaseGStreamer::trackIdFromPadStreamStartOrUniqueID): Deleted.
(WebCore::TrackPrivateBaseGStreamer::trackIdFromStringIdOrIndex): Deleted.
* Source/WebCore/platform/graphics/gstreamer/TrackPrivateBaseGStreamer.h:
(WebCore::TrackPrivateBaseGStreamer::streamId const):
Added, returns m_id. This is different to TrackPrivateBase::id(), since
its implementations first try to return the ID derived from the
container-specific-track-id tag if available.
(WebCore::TrackPrivateBaseGStreamer::gstStreamId const): Added.
(WebCore::TrackPrivateBaseGStreamer::capsChanged):
(WebCore::TrackPrivateBaseGStreamer::stringId const): Deleted.
- Removed m_stringId, instead using m_id everywhere.
- Added m_gstStreamId, storing the stream-id string as reported by
  GStreamer.

* Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.cpp:
(WebCore::VideoTrackPrivateGStreamer::VideoTrackPrivateGStreamer):
(WebCore::VideoTrackPrivateGStreamer::capsChanged):
* Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.h:
Added constructor with an explicit TrackID argument.

* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::~AppendPipeline):
(WebCore::AppendPipeline::didReceiveInitializationSegment):
(WebCore::createOptionalParserForFormat):
(WebCore::AppendPipeline::recycleTrackForPad):
(WebCore::AppendPipeline::linkPadWithTrack):
(WebCore::AppendPipeline::makeWebKitTrack):
Explicitly construct tracks with our IDs, as relying on
TrackPrivateBaseGStreamer to pick one up from stream-id risks collisions.
(WebCore::AppendPipeline::Track::emplaceOptionalParserForFormat):
(WebCore::AppendPipeline::generateTrackId): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.h:
(WebCore::AppendPipeline::Track::Track):
(WebCore::AppendPipeline::tryCreateTrackFromPad):
Instead of generating an ID based on track type and index, parse stream-id,
and add it to the registry so potential ID collisions can be caught and avoided.

* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::filterOutRepeatingTracks):

* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp:
(WebCore::MediaSourcePrivateGStreamer::registerTrackId): Added.
(WebCore::MediaSourcePrivateGStreamer::unregisterTrackId): Added.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.h:
- Added m_trackIdRegistry, to be used by AppendPipeline.

* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourceTrackGStreamer.cpp:
(WebCore::MediaSourceTrackGStreamer::MediaSourceTrackGStreamer):
(WebCore::MediaSourceTrackGStreamer::create):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourceTrackGStreamer.h:

* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
(WebCore::SourceBufferPrivateGStreamer::flush):
(WebCore::SourceBufferPrivateGStreamer::allSamplesInTrackEnqueued):
(WebCore::SourceBufferPrivateGStreamer::precheckInitializationSegment):
(WebCore::SourceBufferPrivateGStreamer::tryRegisterTrackId):
Added, calls `MediaSourcePrivateGStreamer` to try to register an ID.
(WebCore::SourceBufferPrivateGStreamer::tryUnregisterTrackId): ditto
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.h:

* Source/WebCore/platform/graphics/gstreamer/mse/TrackQueue.cpp:
(WebCore::TrackQueue::TrackQueue):
(WebCore::TrackQueue::enqueueObject):
(WebCore::TrackQueue::clear):
(WebCore::TrackQueue::flush):
(WebCore::TrackQueue::notifyWhenLowLevel):
(WebCore::TrackQueue::pop):
(WebCore::TrackQueue::notifyWhenNotEmpty):
(WebCore::TrackQueue::resetNotEmptyHandler):
(WebCore::TrackQueue::checkLowLevel):
* Source/WebCore/platform/graphics/gstreamer/mse/TrackQueue.h:

* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
(WebKitMediaSrcPrivate::streamById): Renamed from streamByName.
(dumpPipeline):
(webKitMediaSrcEmitStreams):
(webKitMediaSrcTearDownStream):
(webKitMediaSrcLoop):
(webKitMediaSrcStreamFlush):
(webKitMediaSrcFlush):
(WebKitMediaSrcPrivate::streamByName): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/286527@main">https://commits.webkit.org/286527@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3da11cbd9c0ff0c1f80a0a5b694d0bb238dcbc9c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76109 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55139 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29011 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80611 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27376 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78225 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64283 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3435 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59676 "Found 1 new test failure: webrtc/utf8-sdp.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17818 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79176 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49568 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65361 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40026 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46964 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22845 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25703 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68096 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23180 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82072 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3481 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2245 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67906 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3635 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65330 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67215 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16783 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11165 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9281 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3429 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6236 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3452 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/6846 "Build was cancelled. Recent messages:OS: Sequoia (15.0.1), Xcode: 16.0") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5210 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -204,6 +205,32 @@ bool getSampleVideoInfo(GstSample* sample, GstVideoInfo& videoInfo)
+    const gchar* streamIdAsCharacters = gst_pad_get_stream_id(pad.get());
+    if (!streamIdAsCharacters)
+        return std::nullopt;
+
```
Might be useful here to check if the pad has a sticky tag event and if the corresponding tag list contains a container-specific-track-id.

vivienne-w replied:
That tag is already handled in https://github.com/WebKit/WebKit/blob/8f96cc81ef906ace167b247fe77a20461e9dc1f4/Source/WebCore/platform/graphics/gstreamer/TrackPrivateBaseGStreamer.cpp#L402 though, and the value is stored in its own `m_trackID`, so I don't think it should be handled here too - these two methods are intended to just replace any uses of `gst_{stream,pad}_get_stream_id`.

Maybe I could name them differently, or add comments to clarify the intention?
(something like `getStreamIdFrom{Pad,Stream}` or `parseGStreamerIdFrom{Pad,Stream}` perhaps)

philn replied:
imho the `getStreamIdFrom{Pad,Stream}` sound more accurate indeed.

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -204,6 +205,32 @@ bool getSampleVideoInfo(GstSample* sample, GstVideoInfo& videoInfo)
+    const gchar* streamIdAsCharacters = gst_stream_get_stream_id(stream.get());
+    if (!streamIdAsCharacters)
+        return std::nullopt;
+
```
Ditto, using `gst_stream_get_tags()`.

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -229,6 +231,48 @@ bool getSampleVideoInfo(GstSample* sample, GstVideoInfo& videoInfo)
 
+std::optional<TrackID> getTrackIdFromPad(const GRefPtr<GstPad>& pad)
+{
+    const gchar* streamIdAsCharacters = gst_pad_get_stream_id(pad.get());
```
This leaks, `gst_pad_get_stream_id()` returns a newly allocated string.

vivienne-w replied:
Thats oddly inconsistent since `gst_stream_get_stream_id()` doesnt, but thanks for pointing it out!

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -229,6 +231,48 @@ bool getSampleVideoInfo(GstSample* sample, GstVideoInfo& videoInfo)
+    if (maybeUUID.has_value())
+        return maybeUUID.value().low();
+
+    size_t position = stringId.find('/');
```
Parsing the stream ID goes against the recommendation from the API docs (see gst_pad_get_stream_id). I think this warrants at least a comment.

philn commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -808,14 +793,20 @@ std::pair<AppendPipeline::CreateTrackResult, AppendPipeline::Track*> AppendPipel
-    GST_DEBUG_OBJECT(pipeline(), "Creating new AppendPipeline::Track with id '%s'", trackStringId.string().utf8().data());
     size_t newTrackIndex = m_tracks.size();
-    m_tracks.append(makeUnique<Track>(trackIndex, trackStringId, streamType, parsedCaps, presentationSize));
+    TrackID preferredTrackId = getTrackIdFromPad(demuxerSrcPad).value_or((TrackID) newTrackIndex);
```
We don't use C casts in WebKit, I think a `reinterpret_cast` might work here?

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
```diff
@@ -37,6 +37,11 @@ namespace WebCore {
 class IntSize;
 class SharedBuffer;
 
+using TrackID = uint64_t;
```
Can the `PRIu64` specifier be used to print those instead of `%lu` ?

vivienne-w committed:
[GStreamer] Use integers as track ID instead of AtomString
https://bugs.webkit.org/show_bug.cgi?id=270638

Reviewed by Philippe Normand.

Switch to using integers for track IDs internally instead of AtomString.

Some notable special cases:
- as MediaStream tracks are recommended by the spec to use UUID strings,
  we just truncate it to 64 bits for the internal TrackID,
  and return the full ID as String (stored in m_gstStreamId) in `trackUID()`
- In MSE, there is the potential for multiple SourceBuffers providing tracks
  with identical IDs. Previously, this was less of an issue since IDs were
  also based on track type, so two tracks "A0" and "V0" would still have
  distinct IDs. However, now they would both get ID 0, and one would get
  dropped in `MediaSourceTrackGStreamer::filterOutRepeatingTracks`
  To work around this, AppendPipeline will register any IDs it creates
  with `MediaSourcePrivateGStreamer::m_trackIdRegistry`, which will
  assign different IDs in case of a collision.
  There has been some consideration as to where to place this registry,
  and how to assign fallback IDs, which is both explained in comments.
  One test that produced these collisions is:
  imported/w3c/web-platform-tests/encrypted-media/clearkey-mp4-playback-temporary-waitingforkey.https.html

* Source/WebCore/platform/graphics/gstreamer/AudioTrackPrivateGStreamer.cpp:
(WebCore::AudioTrackPrivateGStreamer::AudioTrackPrivateGStreamer):
Add constructor with an explicit TrackID argument.
(WebCore::AudioTrackPrivateGStreamer::capsChanged):
* Source/WebCore/platform/graphics/gstreamer/AudioTrackPrivateGStreamer.h:
(WebCore::AudioTrackPrivateGStreamer::create):

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::getTrackIdFromPad): wrapper around gst_pad_get_stream_id, parses stream-id into a TrackID
(WebCore::getTrackIdFromStream): ditto, wraps around gst_stream_get_stream_id
(WebCore::trackIdFromString): parses a stream-id.
Can handle plain integers, the qtdemux-style ".../<ID>" format, and UUIDs.
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:

* Source/WebCore/platform/graphics/gstreamer/InbandTextTrackPrivateGStreamer.cpp:
(WebCore::InbandTextTrackPrivateGStreamer::InbandTextTrackPrivateGStreamer):
Add constructor with an explicit TrackID argument.
* Source/WebCore/platform/graphics/gstreamer/InbandTextTrackPrivateGStreamer.h:
(WebCore::InbandTextTrackPrivateGStreamer::create):

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::notifyPlayerOfTrack):
(WebCore::MediaPlayerPrivateGStreamer::handleTextSample):
(WebCore::MediaPlayerPrivateGStreamer::updateEnabledVideoTrack):
(WebCore::MediaPlayerPrivateGStreamer::updateEnabledAudioTrack):
(WebCore::MediaPlayerPrivateGStreamer::playbin3SendSelectStreamsIfAppropriate):
(WebCore::MediaPlayerPrivateGStreamer::updateTracks):
(WebCore::MediaPlayerPrivateGStreamer::setupCodecProbe):
(WebCore::MediaPlayerPrivateGStreamer::codecForStreamId):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:

* Source/WebCore/platform/graphics/gstreamer/TextSinkGStreamer.cpp:
(webkitTextSinkHandleSample):

* Source/WebCore/platform/graphics/gstreamer/TrackPrivateBaseGStreamer.cpp:
(WebCore::TrackPrivateBaseGStreamer::TrackPrivateBaseGStreamer):
Add constructor with an explicit TrackID argument.
Sets the new m_shouldUsePadStreamId field false, to avoid the ID being changed elsewhere.
(WebCore::TrackPrivateBaseGStreamer::setPad):
(WebCore::TrackPrivateBaseGStreamer::notifyTrackOfStreamChanged):
(WebCore::TrackPrivateBaseGStreamer::installUpdateConfigurationHandlers):
(WebCore::TrackPrivateBaseGStreamer::updateTrackIDFromTags):
(WebCore::TrackPrivateBaseGStreamer::prefixForType): Deleted.
(WebCore::trimStreamId): Deleted.
(WebCore::TrackPrivateBaseGStreamer::generateUniquePlaybin2StreamID): Deleted.
(WebCore::TrackPrivateBaseGStreamer::trackIdFromPadStreamStartOrUniqueID): Deleted.
(WebCore::TrackPrivateBaseGStreamer::trackIdFromStringIdOrIndex): Deleted.
* Source/WebCore/platform/graphics/gstreamer/TrackPrivateBaseGStreamer.h:
(WebCore::TrackPrivateBaseGStreamer::streamId const):
Added, returns m_id. This is different to TrackPrivateBase::id(), since
its implementations first try to return the ID derived from the
container-specific-track-id tag if available.
(WebCore::TrackPrivateBaseGStreamer::gstStreamId const): Added.
(WebCore::TrackPrivateBaseGStreamer::capsChanged):
(WebCore::TrackPrivateBaseGStreamer::stringId const): Deleted.
- Removed m_stringId, instead using m_id everywhere.
- Added m_gstStreamId, storing the stream-id string as reported by
  GStreamer.

* Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.cpp:
(WebCore::VideoTrackPrivateGStreamer::VideoTrackPrivateGStreamer):
(WebCore::VideoTrackPrivateGStreamer::capsChanged):
* Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.h:
Added constructor with an explicit TrackID argument.

* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::~AppendPipeline):
(WebCore::AppendPipeline::didReceiveInitializationSegment):
(WebCore::createOptionalParserForFormat):
(WebCore::AppendPipeline::recycleTrackForPad):
(WebCore::AppendPipeline::linkPadWithTrack):
(WebCore::AppendPipeline::makeWebKitTrack):
Explicitly construct tracks with our IDs, as relying on
TrackPrivateBaseGStreamer to pick one up from stream-id risks collisions.
(WebCore::AppendPipeline::Track::emplaceOptionalParserForFormat):
(WebCore::AppendPipeline::generateTrackId): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.h:
(WebCore::AppendPipeline::Track::Track):
(WebCore::AppendPipeline::tryCreateTrackFromPad):
Instead of generating an ID based on track type and index, parse stream-id,
and add it to the registry so potential ID collisions can be caught and avoided.

* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::filterOutRepeatingTracks):

* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp:
(WebCore::MediaSourcePrivateGStreamer::registerTrackId): Added.
(WebCore::MediaSourcePrivateGStreamer::unregisterTrackId): Added.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.h:
- Added m_trackIdRegistry, to be used by AppendPipeline.

* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourceTrackGStreamer.cpp:
(WebCore::MediaSourceTrackGStreamer::MediaSourceTrackGStreamer):
(WebCore::MediaSourceTrackGStreamer::create):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaSourceTrackGStreamer.h:

* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
(WebCore::SourceBufferPrivateGStreamer::flush):
(WebCore::SourceBufferPrivateGStreamer::allSamplesInTrackEnqueued):
(WebCore::SourceBufferPrivateGStreamer::precheckInitializationSegment):
(WebCore::SourceBufferPrivateGStreamer::tryRegisterTrackId):
Added, calls `MediaSourcePrivateGStreamer` to try to register an ID.
(WebCore::SourceBufferPrivateGStreamer::tryUnregisterTrackId): ditto
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.h:

* Source/WebCore/platform/graphics/gstreamer/mse/TrackQueue.cpp:
(WebCore::TrackQueue::TrackQueue):
(WebCore::TrackQueue::enqueueObject):
(WebCore::TrackQueue::clear):
(WebCore::TrackQueue::flush):
(WebCore::TrackQueue::notifyWhenLowLevel):
(WebCore::TrackQueue::pop):
(WebCore::TrackQueue::notifyWhenNotEmpty):
(WebCore::TrackQueue::resetNotEmptyHandler):
(WebCore::TrackQueue::checkLowLevel):
* Source/WebCore/platform/graphics/gstreamer/mse/TrackQueue.h:

* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp:
(WebKitMediaSrcPrivate::streamById): Renamed from streamByName.
(dumpPipeline):
(webKitMediaSrcEmitStreams):
(webKitMediaSrcTearDownStream):
(webKitMediaSrcLoop):
(webKitMediaSrcStreamFlush):
(webKitMediaSrcFlush):
(WebKitMediaSrcPrivate::streamByName): Deleted.
* Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.h:

Canonical link: https://commits.webkit.org/286527@main
